src/main/scala/net/snowflake/spark/snowflake/io/CloudStorageOperations.scala

```diff
@@ -563,188 +559,158 @@ sealed trait CloudStorage {
-
-    // import java.nio.charset.StandardCharsets
-    // println(new String(data, StandardCharsets.UTF_8))
-
-    if (disable_cache_data) {
-      // The data has been uploaded while reading,
-      // pass through to skip the file upload in this step.
-      val endTime = System.currentTimeMillis()
-      CloudStorageOperations.log.info(
-        s"""${SnowflakeResultSetRDD.WORKER_LOG_PREFIX}:
-           | Finish writing partition ID:$partitionID:
-           | without cache data. Total process time is
-           | ${Utils.getTimeString(endTime - startConvertTime)}
-           | write row count is $rowCount.
-           | Uncompressed data size is ${Utils.getSizeString(dataSize)}.
-           | (Data size = 0 means file is empty, no file is uploaded)
-           |""".stripMargin.filter(_ >= ' '))
-    } else if (readData.isDefined && readData.get.nonEmpty) {
-      // Upload the file with retry and backoff.
-      // default maxRetryCount is 10 which is configurable.
-      var retryCount = 0
-      var error: Option[Exception] = None
-      var uploadDone = false
-      var startUploadTime: Long = 0
-      val data = readData.get
-      dataSize = data.size
-      do {
-        try {
-          startUploadTime = System.currentTimeMillis()
-
-          if (storageInfo.isDefined) {
-            // Update data with StorageInfo
-            val uploadStream =
-              createUploadStream(fileName, Some(directory), compress, storageInfo.get)
-            uploadStream.write(data)
-            uploadStream.close()
-          } else if (fileTransferMetadata.isDefined) {
-            // Set up proxy info if it is configured.
-            val proxyProperties = new Properties()
-            proxyInfo match {
-              case Some(proxyInfoValue) =>
-                proxyInfoValue.setProxyForJDBC(proxyProperties)
-              case None =>
-            }
-
-            // Update data with FileTransferMetadata
-            val inStream = new ByteArrayInputStream(data)
-            SnowflakeFileTransferAgent.uploadWithoutConnection(
-              SnowflakeFileTransferConfig.Builder.newInstance()
-                .setSnowflakeFileTransferMetadata(fileTransferMetadata.get)
-                .setUploadStream(inStream)
-                .setRequireCompress(compress)
-                .setOcspMode(OCSPMode.FAIL_OPEN)
-                .setProxyProperties(proxyProperties)
-                .build())
-          }
-          val endUploadTime = System.currentTimeMillis()
-
-          TestHook.raiseExceptionIfTestFlagEnabled(
-            TestHookFlag.TH_GCS_UPLOAD_RAISE_EXCEPTION,
-            "Negative test to raise error when uploading data to GCS"
-          )
+    try {
+      // Read data and upload to cloud storage
+      doUploadPartition(rows, format, compress, directory, partitionID,
+                        storageInfo, fileTransferMetadata)
+    } catch {
+      // Hit exception when uploading the file
+      case e: Exception => {
```

Comment:
Do we want to do this on all exceptions or just throttling exceptions ?

Reply:
We do this for all exceptions.

Reply:
Ideally, we should sleep only when throttling issue happens. But it is difficult to achieve because there are three kinds of cloud services and each one may raise multiple kinds of exceptions. And it doesn't deserve to spend much time here.

Reply:
In the case of Nielsen, what specific exception was raised ? My concern is that this is too broad and screws up some fail-fast cases. 
If you do capture errors in such a broad manner, we should also have some param to turn off retries if needed.

Reply:
I agree, it seems that catching all exceptions is too general. We would incur extra, unnecessary sleep latency for other errors as well.

Reply:
I remember two kinds of exception is hit for Nielsen. One is something like "Sever is busy". I don't remember the other. Some month ago, another customer hit throttling issue on AWS, the error is only generic TIMEOUT.

By default, it retries 4 times. The average extra sleep time is (1.5 + 3 + 6 + 9) * 0.75 = 16 seconds.
Currently, for fail-fast cases, they configure spark spark.task.maxFailures as 1.It sleeps about 1 second.
It sounds the side-affect is not a big problem.

The retry count is controlled by spark, I think it may be not necessary to introduce new para on spark connector side.
But we can add a param to disable the extra sleep. Do you think it is necessary?

Reply:
I agree that it is pretty general to catch all exceptions. The another thing we need to consider is the task failure possibility. According to the OOB telemetry messages, only 2 upload errors are seen (in past 30 days). It can prove the task failure case is rare though the OOB message are sent only by SC 2.7.2 and 2.8.0.